[Gekidou] Add command autocomplete

larkox opened the pull request:
#### Summary
Add command Autocomplete, including Apps commands

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-39682

#### Release Note
```release-note
NONE
```

larkox commented on app/components/autocomplete/autocomplete.tsx:
```diff
@@ -192,14 +197,18 @@ const Autocomplete = ({
                     }
-                    {/* <SlashSuggestion
+                    {showCommands &&
+                    <SlashSuggestion
```
Unmounting this component like this breaks the command cache. We can store the command cache somewhere else and not in the component, but then we have to handle how long do we trust the cache, etc... (it was previously done in redux). 0/5 whether we should worry about this, since this is for uncommon cases.

larkox replied:
In memory singleton were you ask for the commands for certain team, if the cache is "fresh" get it from the cache, if not, it ask the server. It seems fairly simple. Will do.

Regarding how uncommon, now that I think about it, may be more common than I though. Any moment that while you are writing a command you use at-mention, channel-mentions, or emojis, it will unmount the component, and then has to mount it again and fetch again.

larkox commented on app/components/autocomplete/slash_suggestion/app_command_parser/app_command_parser.ts:
```diff
@@ -0,0 +1,1437 @@
+    formatMessage(config: {id: string; defaultMessage: string}, values?: {[name: string]: any}): string;
+}
+
+// TODO: Implemnet app bindings
```
Ticket to be created before merging the PR.

larkox replied:
This file is from the integrations team. This file is also shared with webapp. Therefore, any changes here may cause trouble for them, since they would have to do the same changes in webapp.

@catalintomai heads up that some changes here are happening (because of the new architecture) and that maybe your team will have to look into this for V2 beta.

larkox commented on app/components/autocomplete/slash_suggestion/app_command_parser/app_command_parser.ts:
```diff
@@ -0,0 +1,1437 @@
+                        break;
+                    case AppFieldTypes.USER: {
+                        const userID = (f.value as AppSelectOption).value;
+                        let user: UserModel | UserProfile = (await queryUsersById(this.database, [userID]))[0];
+                        if (!user) {
+                            const res = await fetchUsersByIds(this.serverUrl, [userID]);
+                            if ('error' in res) {
+                            // Silently fail on default value
+                                break;
+                            }
+                            user = res.users[0] || res.existingUsers[0];
+                        }
```
This pattern repeats several places in this file (also for channels). The idea is "if the user is not in local, get it from server". The question is... do we want to store the user/channel in local (fetchOnly = [true | false]?)? Reasons not to do it? If we don't belong to that channel, we won't receive updates for it, so we will have stale data.

larkox replied:
For example:
- I use a command with channel autocomplete, and run the command with ChannelA
- I don't have ChannelA on my database because I am not a member, so I fetch it from the server and store it.
- Someone changes the display name of ChannelA
- I don't receive any websocket event because I am no longer part of the channel.
- If I look for any reason into the information in the database, the channel information will be old (the old display name). Also, not sure if there are any methods to remove it from the database.

I think that right now it is not a big deal in terms of old data, since whenever we look for channels where we do not belong, we ask the server, we do not look the database, but still, may be a problem.

larkox commented on app/components/autocomplete/slash_suggestion/slash_suggestion_item.tsx:
```diff
@@ -0,0 +1,171 @@
+        );
+    } else if (icon.startsWith('data:')) {
+        if (icon.startsWith('data:image/svg+xml')) {
+            // TODO: What base64 library should we use? Security implications on doing things like this?
```
We used to use a base64 library on V1 that is not yet installed in V2. Just wanted to make sure if we want to install that one, install another one (or use something in one already installed), or if we want to handle this differently.

Also, since this information comes from the plugin, it may be considered "unsafe string" and we may have to do some extra checks here (or trust SvgXml to do the checks for us).

enahum replied:
I believe we have `base64-js` already in `node_modules` does that one work for you?

larkox replied:
After looking a bit deeper, in theory won't be needed. There is a `atob` function deprecated, and it recommend to use `Buffer.from`. I am using that in the latest commit. Hopefully is good enough.

larkox commented on types/api/apps.d.ts:
```diff
@@ -178,11 +178,11 @@ type AppField = {
-    description?: string;
-    hint?: string;
-    iconData?: string;
+    Suggestion: string;
```
This is the shape it has when it arrives from the server.

larkox commented:
Another outstanding issue is considered here: https://community.mattermost.com/core/pl/ydk9nnff8bg1ukf1kdk5j55mcc

TL;DR: Behaviour of commands when the cursor is in a different place from the end.

avinashlng1080 reviewed (changes requested):
I haven't yet completed the review but I'm aiming to complete it today itself.

enahum reviewed (approved):
I wonder if we have the bang in the compass icons instead of using the `slash_command_error.png` files. cc: @matthewbirtch

matthewbirtch commented:
> I wonder if we have the bang in the compass icons instead of using the `slash_command_error.png` files. cc: @matthewbirtch

@enahum I'm not 100% clear of the context of this icon, but maybe we can use the `alert-outline` icon if it represents an error.
<img width="222" alt="image" src="https://user-images.githubusercontent.com/2040554/159557189-b51e2bc0-7ccc-4650-8b40-f7a74204e7bb.png">

larkox commented:
@matthewbirtch Here is an example of were it is shown (in V1):
https://user-images.githubusercontent.com/35588783/129263391-491f5f8b-7b9f-41db-9e25-afc5f5f1d41b.png

larkox commented on app/components/autocomplete/slash_suggestion/app_slash_suggestion/app_slash_suggestion.tsx:
```diff
@@ -0,0 +1,211 @@
+    const serverUrl = useServerUrl();
+    const appCommandParser = useRef<AppCommandParser>(new AppCommandParser(serverUrl, intl, channelId, currentTeamId, rootId, theme));
+    const [dataSource, setDataSource] = useState<AutocompleteSuggestion[]>(emptySuggestonList);
+    const active = isAppsEnabled && Boolean(dataSource.length);
```
App suggestion don't do unneeded server requests, start from the beginning of the line, and go to the end, so no need to stop.

larkox commented on app/components/autocomplete/slash_suggestion/slash_suggestion.tsx:
```diff
@@ -0,0 +1,236 @@
+    const [dataSource, setDataSource] = useState<AutocompleteSuggestion[]>(emptySuggestionList);
+    const [commands, setCommands] = useState<Command[]>();
+
+    const active = Boolean(dataSource.length);
```
Here we could use the stop. Will implement.

matthewbirtch commented:
> @matthewbirtch Here is an example of were it is shown (in V1): https://user-images.githubusercontent.com/35588783/129263391-491f5f8b-7b9f-41db-9e25-afc5f5f1d41b.png

Okay, it should be fine to use either of these:

![image](https://user-images.githubusercontent.com/2040554/159732685-1edcb148-90ed-47c1-b8d7-4431a615ec49.png)

![image](https://user-images.githubusercontent.com/2040554/159732762-ac9b5398-5873-47d2-a5bf-e46233ee93a0.png)

larkox commented:
@matthewbirtch Would this look ok like this?
(I did some tweak on the code so all commands show this icon, but in the real world it would only show on errors)
![Screenshot from 2022-03-23 17-31-38](https://user-images.githubusercontent.com/1933730/159749558-1d215a96-00f3-4118-93e4-468f2ace938c.png)

avinashlng1080 reviewed (approved):
LGTM.  I'm going to approve this PR so as not to block it.    I definitely need to spend more time understanding this feature.  Thanks a lot @larkox  for the call; it provided a lot of context.

enahum commented:
@larkox some conflicts and more importantly the Buffer thing

larkox commented:
I added a Integrations Manager to handle the cache in memory (ephemeral storage). I also added some of the functions used in the App Command Parser and the trigger id, since this is where it will probably live. This should be finished when all the Apps and integrations work is in place.

Regarding Buffer, you were right. Neither Buffer not btoa are defined in ReactNative. I installed the package we are using on V1.
